src/compiler/reverse.jl

```diff
@@ -26,13 +26,28 @@ end
 unwrapquote(x) = x
 unwrapquote(x::QuoteNode) = x.value
 
-is_literal_getproperty(ex) =
-  (iscall(ex, Base, :getproperty) || iscall(ex, Core, :getfield) || iscall(ex, Base, :getfield)) &&
-  ex.args[3] isa Union{QuoteNode,Integer}
+is_getproperty(ex) = iscall(ex, Base, :getproperty)
 
 function instrument_getproperty!(ir, v, ex)
-  is_literal_getproperty(ex) ?
-    (ir[v] = xcall(Zygote, :literal_getproperty, ex.args[2], Val(unwrapquote(ex.args[3])))) :
+  if is_getproperty(ex)
+    if ex.args[3] isa Union{QuoteNode,Integer}
+      ir[v] = xcall(Zygote, :literal_getproperty, ex.args[2], Val(unwrapquote(ex.args[3])))
+    else
+      f = insert!(ir, v, :(Val($(ex.args[3]))))
+      ir[v] = xcall(Zygote, :literal_getproperty, ex.args[2], f)
```

Comment:
This is what the fallback pullback definition for `getproperty` was doing anyways and it doesn't look like anyone is defining adjoints for `getproperty` explicitely: https://juliahub.com/ui/RepoSearch?q=%40adjoint%20%28Base%5C.%29%3Fgetproperty&r=true

Reply:
It is and inconsistent to have literall getproperty do the lifting things to `Val` in the instrumentation call, but to have `literal_getfield` define a fallback method.
We should do them both the same.
and maybe we should use a little `@eval` loop

```
for getfunc in (:getfield, :getproperty) 
    function $(Symbol(:instrument_, getfunc, :!)(ir, v, ex)
        if $(Symbol(:is_, getfunc) 
```

Reply:
Why do we need to refactor like that? Seems like doing it this way makes the intention muddier and still removes the possibility where an adjoint over `getproperty` might be needed in the future.

Reply:
I don't really have a strong opinion here. For `getproperty` this change was necessary, because we want to allow people to define custom adjoints using `literal_getproperty`, but recurse into `getproperty` otherwise. This should happen no matter whether the second argument is a literal or not, so we always need to instrument the IR accordingly. For `getfield`, this way is fine, since it's not a generic function and no user should define an adjoint for `literal_getfield`.

Reply:
How so?

Reply:
Are you asking why users shouldn't define an adjoint for `literal_getfield`?

Reply:
oh sorry no it was github muddying the order of text, I was asking what Dhairya ment but you comment made it clear

Now I mostly get it.
We should include comments in the code that explain that users should not be able to overload `literal_getfield` but that we do need to allow that for `literal_getproperty`, so that that is not lost.

Reply:
The general problem here is is that design with `literal_getproperty` in some ways was inherently flawed, because for `getproperty` it was intended that `_pullback` falls back to `literal_getproperty`, but we actually want the opposite to happen, since Zygote should fall back to recursing into the `getproperty` implementation by default.

Reply:
>  We should include comments in the code that explain that users should not be able to overload literal_getfield but that we do need to allow that for literal_getproperty, so that that is not lost.

Ok, I can add some comments that explain this a bit better.